Source/Core/DolphinQt2/FIFOPlayerWindow.cpp

```diff
@@ -0,0 +1,348 @@
+
+  setLayout(layout);
+}
+
+void FIFOPlayerWindow::ConnectWidgets()
+{
+  connect(m_load, &QPushButton::pressed, this, &FIFOPlayerWindow::LoadRecording);
+  connect(m_save, &QPushButton::pressed, this, &FIFOPlayerWindow::SaveRecording);
+  connect(m_record, &QPushButton::pressed, this, &FIFOPlayerWindow::ToggleRecording);
+  connect(m_button_box, &QDialogButtonBox::rejected, this, &FIFOPlayerWindow::reject);
+  connect(m_early_memory_updates, &QCheckBox::toggled, this,
+          &FIFOPlayerWindow::OnEarlyMemoryUpdatesChanged);
+  connect(m_frame_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnFrameFromChanged);
+  connect(m_frame_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnFrameToChanged);
+
+  connect(m_object_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnObjectFromChanged);
+  connect(m_object_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnObjectToChanged);
+}
+
+void FIFOPlayerWindow::LoadRecording()
+{
+  QString path = QFileDialog::getOpenFileName(this, tr("Open FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  emit LoadFIFO(path);
+}
+
+void FIFOPlayerWindow::SaveRecording()
+{
+  QString path = QFileDialog::getSaveFileName(this, tr("Save FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOPlayerWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!recorder.IsRecording())
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(),
+                            [this] { QueueOnObject(this, [this] { OnRecordingDone(); }); });
+
+    m_save->setEnabled(false);
+    m_frame_range_to->setEnabled(false);
+    m_frame_range_from->setEnabled(false);
+    m_object_range_from->setEnabled(false);
+    m_object_range_to->setEnabled(false);
+    m_early_memory_updates->setEnabled(false);
+    m_frame_record_count->setEnabled(false);
+
+    m_record->setText(tr("Stop"));
+  }
+  else
+  {
+    recorder.StopRecording();
+  }
+
+  UpdateInfo();
+}
+
+void FIFOPlayerWindow::OnEmulationStarted()
+{
+  m_record->setEnabled(FifoPlayer::GetInstance().GetFile() == nullptr);
+  m_load->setEnabled(false);
+}
+
+void FIFOPlayerWindow::OnEmulationStopped()
+{
+  m_record->setEnabled(false);
+  m_load->setEnabled(true);
+
+  // If we have previously been recording, stop now.
+  if (FifoRecorder::GetInstance().IsRecording())
```

Comment:
Following this logic is a bit weird: we check if the FifoRecorder is recording, then call ToggleRecording, which changes its behavior depending on if FifoRecorder is recording. It would be simpler to call the required logic ("StopRecording"?) directly.

The dual-behavior ToggleRecording function is required because the "Record" button sometimes masquerades as a "Stop" button: this would be simpler to reason about if there was a separate "Stop" button that changed visibility along with "Record", in which case each button could be connected directly to whichever function it needed to call (e.g. "StartRecording", "StopRecording").